include/itkImageToTorchImageFilter.hxx

```diff
@@ -18,48 +18,76 @@
+::GenerateInputRequestedRegion()
+{
+  // Is this correct?!!!  Instead, is the superclass implementation sufficient?!!!
+  // Call the superclass's implementation of this method
+  Superclass::GenerateInputRequestedRegion();
+
+  // Get pointers to the input image.
+  typename Self::InputImageType *inputImage = const_cast< typename Self::InputImageType * >( this->GetInput( 0 ) );
+
+  if( inputImage != nullptr )
+    {
+    inputImage->SetRequestedRegionToLargestPossibleRegion();
+    }
+}
 
 
-template <typename TInputImage, typename TOutputImage>
+template< typename TInputImage, typename TOutputTorchImage >
 void
-ImageToTorchImageFilter<TInputImage, TOutputImage>
-::PrintSelf(std::ostream & os, Indent indent) const
+ImageToTorchImageFilter< TInputImage, TOutputTorchImage >
+::BeforeThreadedGenerateData()
 {
-  Superclass::PrintSelf(os, indent);
+  // Call the superclass' implementation of this method
+  Superclass::BeforeThreadedGenerateData();
+
+  typename Self::InputImageType const * const inputImage = this->GetInput();
+  const typename Self::InputImageSizeType inputSize = inputImage->GetBufferedRegion().GetSize();
+  typename Self::InputImagePixelType const * const inputBufferPointer = inputImage->GetBufferPointer();
+
+  typename Self::OutputTorchImageType * const outputTorchImage = this->GetOutput();
+  const typename Self::OutputTorchImageSizeType outputSize = outputTorchImage->GetBufferedRegion().GetSize();
+
+  itkAssertOrThrowMacro( inputSize == outputSize, "ImageToTorchImageFilter: input and output images must have the same size in each dimension");
+
+  // Do the memory copy of the whole image.
+  // Do we need to Allocate() the output?!!!
+
+  torch::Tensor inputTensor =
+    torch::from_blob( ( void * )inputBufferPointer, outputTorchImage->m_Tensor.sizes(), outputTorchImage->m_Tensor.options().device( torch::kCPU ) );
+  outputTorchImage->m_Tensor = inputTensor.to( outputTorchImage->m_Tensor.options() );
 }
 
 
-template <typename TInputImage, typename TOutputImage>
+template< typename TInputImage, typename TOutputTorchImage >
 void
-ImageToTorchImageFilter<TInputImage, TOutputImage>
-::DynamicThreadedGenerateData(const OutputRegionType & outputRegion)
+ImageToTorchImageFilter< TInputImage, TOutputTorchImage >
+::DynamicThreadedGenerateData( const typename Self::OutputTorchImageRegionType &itkNotUsed( outputRegion ) )
 {
-  OutputImageType *      output = this->GetOutput();
-  const InputImageType * input = this->GetInput();
-  using InputRegionType = typename InputImageType::RegionType;
-  InputRegionType inputRegion = InputRegionType(outputRegion.GetSize());
-
-  itk::ImageRegionConstIterator<InputImageType> in(input, inputRegion);
-  itk::ImageRegionIterator<OutputImageType>     out(output, outputRegion);
-
-  for (in.GoToBegin(), out.GoToBegin(); !in.IsAtEnd() && !out.IsAtEnd(); ++in, ++out)
-  {
-    out.Set(in.Get());
-  }
+  // We did the memory copy already.  Nothing to do here.
```

Comment:
Then maybe go the route of overriding `GenerateData()`. Less method definitions. Mainly rename the current `BeforeThreadedGenerateData()` into `GenerateData()`. See what the superclass [GenerateData()](https://github.com/InsightSoftwareConsortium/ITK/blob/v5.1.1/Modules/Core/Common/include/itkImageSource.hxx#L210-L245) does. I guess this class will only need GPU allocate and copy.